File: processing/roundabout_review.py
"""Pure roundabout candidate/review normalization and join helpers."""

from __future__ import annotations

import math
from typing import Any

SCHEMA_VERSION = 1
VALID_STATUSES = {"accepted", "rejected"}


def valid_candidate_geometry(candidate: dict[str, Any]) -> bool:
    center = candidate.get("center") or {}
    lat = center.get("lat")
    lng = center.get("lng")
    if not isinstance(lat, (int, float)) or not isinstance(lng, (int, float)):
        return False
    if not math.isfinite(lat) or not math.isfinite(lng):
        return False
    classification = candidate.get("classification")
    if classification == "mini_roundabout":
        return isinstance(candidate.get("radiusM"), (int, float)) and candidate["radiusM"] > 0
    paths = candidate.get("paths")
    if not isinstance(paths, list) or not paths:
        return False
    for path in paths:
        if not isinstance(path, list) or len(path) < 2:
            continue
        if all(
            isinstance(coord, list)
            and len(coord) >= 2
            and isinstance(coord[0], (int, float))
            and isinstance(coord[1], (int, float))
            and math.isfinite(coord[0])
            and math.isfinite(coord[1])
            for coord in path
        ):
            return True
    return False


def join_roundabout_reviews(
    candidates_payload: dict[str, Any],
    review_data: dict[str, Any] | None,
) -> dict[str, Any]:
    blocking_issues: list[dict[str, Any]] = []
    if candidates_payload.get("schemaVersion") != SCHEMA_VERSION:
        blocking_issues.append({"code": "invalid_candidate_schema"})
    reviews_payload = review_data if isinstance(review_data, dict) else {}
    if reviews_payload.get("schemaVersion") != SCHEMA_VERSION:
        blocking_issues.append({"code": "invalid_review_schema"})
    reviews = reviews_payload.get("reviews")
    if not isinstance(reviews, dict):
        reviews = {}
        blocking_issues.append({"code": "invalid_reviews_map"})

    candidates = candidates_payload.get("roundabouts")
    if not isinstance(candidates, list):
        candidates = []
        blocking_issues.append({"code": "invalid_candidates_list"})

    seen: set[str] = set()
    accepted = []
    rejected = []
    pending = []
    stale = []
    items = []

    for candidate in candidates:
        if not isinstance(candidate, dict):
            blocking_issues.append({"code": "invalid_candidate"})
            continue
        candidate_id = candidate.get("id")
        fingerprint = candidate.get("fingerprint")
        if not isinstance(candidate_id, str) or not candidate_id or not isinstance(fingerprint, str):
            blocking_issues.append({"code": "invalid_candidate_identity", "id": candidate_id})
            continue
        if candidate_id in seen:
            blocking_issues.append({"code": "duplicate_candidate_id", "id": candidate_id})
            continue
        seen.add(candidate_id)
        review = reviews.get(candidate_id)
        state = "pending"
        if review is not None:
            if not isinstance(review, dict) or review.get("status") not in VALID_STATUSES:
                blocking_issues.append({"code": "invalid_review", "id": candidate_id})
            elif review.get("fingerprint") != fingerprint:
                state = "stale"
            else:
                state = review["status"]
        item = {"candidate": candidate, "review": review, "state": state}
        items.append(item)
        if state == "accepted":
            if not valid_candidate_geometry(candidate):
                blocking_issues.append({"code": "invalid_accepted_geometry", "id": candidate_id})
            accepted.append(candidate)
        elif state == "rejected":
            rejected.append(candidate)
        elif state == "stale":
            stale.append(candidate)
        else:
            pending.append(candidate)

    orphaned = [
        {"id": key, "review": value}
        for key, value in sorted(reviews.items())
        if key not in seen
    ]
    warnings = []
    coverage = candidates_payload.get("coverage") or {}
    if coverage.get("miniRoundaboutNodes") != "available":
        warnings.append({"code": "mini_roundabout_coverage_incomplete"})
    for candidate in candidates:
        if isinstance(candidate, dict) and candidate.get("warnings"):
            warnings.append({"code": "candidate_warnings", "id": candidate.get("id")})

    if pending:
        blocking_issues.append({"code": "pending_reviews", "count": len(pending)})
    if stale:
        blocking_issues.append({"code": "stale_reviews", "count": len(stale)})

    return {
        "accepted": accepted,
        "rejected": rejected,
        "pending": pending,
        "stale": stale,
        "orphaned": orphaned,
        "items": items,
        "warnings": warnings,
        "blockingIssues": blocking_issues,
        "summary": {
            "total": len(items),
            "accepted": len(accepted),
            "rejected": len(rejected),
            "pending": len(pending),
            "stale": len(stale),
            "orphaned": len(orphaned),
            "warnings": len(warnings),
        },
        "coverage": coverage,
    }

File: editor/lib/roundaboutReview.mjs
export const ROUNDABOUT_REVIEW_STATUSES = new Set(["accepted", "rejected"]);

function validGeometry(candidate) {
  const lat = Number(candidate?.center?.lat);
  const lng = Number(candidate?.center?.lng);
  if (!Number.isFinite(lat) || !Number.isFinite(lng)) return false;
  if (candidate?.classification === "mini_roundabout") {
    return Number(candidate?.radiusM) > 0;
  }
  return Array.isArray(candidate?.paths)
    && candidate.paths.some((path) =>
      Array.isArray(path)
      && path.length >= 2
      && path.every((coord) =>
        Array.isArray(coord)
        && coord.length >= 2
        && Number.isFinite(Number(coord[0]))
        && Number.isFinite(Number(coord[1])),
      ));
}

export function joinRoundaboutReviews(candidatesPayload = {}, reviewData = {}) {
  const blockingIssues = [];
  if (candidatesPayload?.schemaVersion !== 1) blockingIssues.push({ code: "invalid_candidate_schema" });
  if (reviewData?.schemaVersion !== 1) blockingIssues.push({ code: "invalid_review_schema" });
  const reviews = reviewData?.reviews && typeof reviewData.reviews === "object" && !Array.isArray(reviewData.reviews)
    ? reviewData.reviews
    : {};
  if (reviews !== reviewData?.reviews) blockingIssues.push({ code: "invalid_reviews_map" });
  const candidates = Array.isArray(candidatesPayload?.roundabouts) ? candidatesPayload.roundabouts : [];
  if (!Array.isArray(candidatesPayload?.roundabouts)) blockingIssues.push({ code: "invalid_candidates_list" });

  const seen = new Set();
  const result = { accepted: [], rejected: [], pending: [], stale: [], orphaned: [], items: [] };
  for (const candidate of candidates) {
    const id = candidate?.id;
    if (typeof id !== "string" || !id || typeof candidate?.fingerprint !== "string") {
      blockingIssues.push({ code: "invalid_candidate_identity", id });
      continue;
    }
    if (seen.has(id)) {
      blockingIssues.push({ code: "duplicate_candidate_id", id });
      continue;
    }
    seen.add(id);
    const review = reviews[id];
    let state = "pending";
    if (review !== undefined) {
      if (!review || typeof review !== "object" || !ROUNDABOUT_REVIEW_STATUSES.has(review.status)) {
        blockingIssues.push({ code: "invalid_review", id });
      } else if (review.fingerprint !== candidate.fingerprint) {
        state = "stale";
      } else {
        state = review.status;
      }
    }
    result.items.push({ candidate, review: review ?? null, state });
    result[state].push(candidate);
    if (state === "accepted" && !validGeometry(candidate)) {
      blockingIssues.push({ code: "invalid_accepted_geometry", id });
    }
  }
  result.orphaned = Object.entries(reviews)
    .filter(([id]) => !seen.has(id))
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([id, review]) => ({ id, review }));
  const warnings = [];
  if (candidatesPayload?.coverage?.miniRoundaboutNodes !== "available") {
    warnings.push({ code: "mini_roundabout_coverage_incomplete" });
  }
  for (const candidate of candidates) {
    if (Array.isArray(candidate?.warnings) && candidate.warnings.length) {
      warnings.push({ code: "candidate_warnings", id: candidate.id });
    }
  }
  if (result.pending.length) blockingIssues.push({ code: "pending_reviews", count: result.pending.length });
  if (result.stale.length) blockingIssues.push({ code: "stale_reviews", count: result.stale.length });
  return {
    ...result,
    warnings,
    blockingIssues,
    coverage: candidatesPayload?.coverage || {},
    summary: {
      total: result.items.length,
      accepted: result.accepted.length,
      rejected: result.rejected.length,
      pending: result.pending.length,
      stale: result.stale.length,
      orphaned: result.orphaned.length,
      warnings: warnings.length,
    },
  };
}

function circleCoordinates(center, radiusM, steps = 32) {
  const lat = Number(center?.lat);
  const lng = Number(center?.lng);
  const latRadius = radiusM / 111_320;
  const lngRadius = latRadius / Math.max(0.1, Math.cos((lat * Math.PI) / 180));
  const ring = [];
  for (let index = 0; index <= steps; index += 1) {
    const angle = (index / steps) * Math.PI * 2;
    ring.push([lng + Math.cos(angle) * lngRadius, lat + Math.sin(angle) * latRadius]);
  }
  return ring;
}

export function roundaboutReviewGeoJson(joined) {
  const lines = [];
  const points = [];
  const corridors = [];
  for (const item of joined?.items || []) {
    const candidate = item.candidate;
    const properties = {
      id: candidate.id,
      state: item.state,
      classification: candidate.classification,
      warning: Boolean(candidate.warnings?.length),
    };
    if (candidate.classification === "mini_roundabout") {
      points.push({ type: "Feature", geometry: { type: "Point", coordinates: [candidate.center.lng, candidate.center.lat] }, properties });
      corridors.push({ type: "Feature", geometry: { type: "Polygon", coordinates: [circleCoordinates(candidate.center, Number(candidate.radiusM) || 10)] }, properties });
    } else {
      for (const path of candidate.paths || []) {
        const coordinates = path.map(([lat, lng]) => [lng, lat]);
        if (coordinates.length < 2) continue;
        lines.push({ type: "Feature", geometry: { type: "LineString", coordinates }, properties });
      }
    }
  }
  return {
    lines: { type: "FeatureCollection", features: lines },
    points: { type: "FeatureCollection", features: points },
    corridors: { type: "FeatureCollection", features: corridors },
  };
}

export function filterRoundaboutItems(items, filter = "all") {
  if (filter === "all") return [...(items || [])];
  if (filter === "warnings") return (items || []).filter((item) => item.candidate?.warnings?.length);
  return (items || []).filter((item) => item.state === filter);
}
